demos/adjoint/adjoint_demo.py

```diff
@@ -0,0 +1,342 @@
+# - $T_{\text{average}}$ is the average temperature profile representing mantle's geotherm.
+# - $T_{F}$ is the the temperature field at final time-step.
+# - $T_{\text{obs}}$ is the observed temperature field at the final time-step.
+# - $u_{\text{obs}}$ is the observed velocity field at *each time-step*.
+# - $\alpha_u$, $\alpha_d$, $\alpha_s$ are the three different
+#   weighting terms for the velocity, damping and smoothing terms.
+#
+# We define the objective functional as
+# $$ \text{Objective Functional}= \int_{\Omega}(T - T_{\text{obs}}) ^ 2 \, dx \\
+#                  +\alpha_u\, \frac{D_{T_{obs}}}{N\times D_{u_{obs}}}\sum_{i}\int_{\partial \Omega_{\text{top}}}(u - u_{\text{obs}}) \cdot(u - u_{\text{obs}}) \, ds \\
+#                  +\alpha_s\, \frac{D_{T_{obs}}}{D_{\text{smoothing}}}\int_{\Omega} \nabla(T_{ic} - T_{\text{average}}) \cdot \nabla(T_{ic} - T_{\text{average}}) \, dx \\
+#                  +\alpha_d\, \frac{D_{T_{obs}}}{D_{\text{damping}}}\int_{\Omega}(T_{ic} - T_{\text{average}}) ^ 2 \, dx $$
+
+# With the three *normlisation terms* of:
+# + $D_{\text{damping}} = \int_{\Omega} T_{\text{average}}^2 \, dx$,
+# + $D_{\text{smoothing}} = \int_{\Omega} \nabla T_{\text{obs}} \cdot \nabla T_{\text{obs}} \, dx$,
+# + $D_{T_{obs}} = \int_{\Omega} T_{\text{obs}} ^ 2 \, dx$, and
+# + $D_{\text{damping}} = \int_{\partial \Omega_{\text{top}}} u_{\text{obs}} \cdot u_{\text{obs}} \, ds$
+#
+# which we define with the following definitions of `objective`
+
+# +
+# Define the component terms of the overall objective functional
+damping = assemble((Tic - Taverage) ** 2 * dx)
+norm_damping = assemble(Taverage**2 * dx)
+smoothing = assemble(dot(grad(Tic - Taverage), grad(Tic - Taverage)) * dx)
+norm_smoothing = assemble(dot(grad(Tobs), grad(Tobs)) * dx)
+norm_obs = assemble(Tobs**2 * dx)
+norm_u_surface = assemble(dot(uobs, uobs) * ds_t)
+
+# Temperature misfit between solution and observation
+t_misfit = assemble((T - Tobs) ** 2 * dx)
+
+# Weighting terms
+alpha_u = 1e-1
+alpha_d = 1e-2
+alpha_s = 1e-1
+
+# Define the overall objective functional
+objective = (
+    t_misfit +
+    alpha_u * (norm_obs * u_misfit / timesteps / norm_u_surface) +
+    alpha_d * (norm_obs * damping / norm_damping) +
+    alpha_s * (norm_obs * smoothing / norm_smoothing)
+)
+# -
+
+# Defining the Reduced Functional and Optimisation Method
+# ===============================
+#
+# In optimisation terminology, a reduced functional is a functional that takes a given value for the control and outputs the value of the objective functional defined for it. It does this without explicitly depending on all the intermediary state variables, hence the name "reduced".
+#
+# To define the reduced functional, we provide the class with an objective (which is an overloaded UFL object) and the control. Both of these are essential for formulating the reduced functional.
+
+# +
+# Define the object for pyadjoint
+reduced_functional = ReducedFunctional(objective, control)
+# -
```

Comment:
```suggestion
reduced_functional = ReducedFunctional(objective, control)
```

I don't think comments like this are necessary with the big description above. Plus, it's clearer not to explicitly mark small blocks like this as code.